### ISSUE TITLE
NVDA Not Reading Alt Text for MathML in Firefox

### ISSUE THREAD

**Issue by RD-Brown** [nvaccess/nvda#5665]
I work for Pearson Education’s accessibility team, and we have identified an issue where NVDA does not read alt text descriptions for MathML content used in online assessments with the Firefox browser.  We are using alt text descriptions because accurate reading of MathML content varies so widely depending on screen reader versions and the browser being used.
(XML attached.)

[NVDA_BugExample.txt](https://github.com/nvaccess/nvda/files/81757/NVDA_BugExample.txt)
1.	Open the following link in Firefox:  uat-collegeboard.testnav.com/client/index.html#login?username=LGN557120611&password=MDYQWJ8G
2.	On the Start Test Page, press B until the Start Test Now button is selected, and press Enter.
3.	On the Section Start Page, press B until the Section Start button is selected, and press Enter.
4.	On the Question Page, press H until the Question heading is selected.
5.	Press Down arrow.

•	Expected:  NVDA should read the following text:
It is given that sine x equals, a where x is the radian measure of an angle and
•	Actual:  NVDA skips the MathML alt text and instead reads:
It is given that   where x is the radian measure of an angle and
•	Expected:  On the following line, NVDA should read:
the fraction pi over 2, is less than x, and x is less than pi.
•	Actual:  NVDA reads nothing.

Interestingly, MathML is read in the radio button choices shown on this page, but again not the alt text which was coded for these items.  Instead, the MathML is being read directly.

**Comment by derekriemer** [comment]
I believe NVDA purposely ignores alt-text because it renders Mathml on 
its own. Since the alt-text is not able to be presented in both nemeth 
and speech it presents a problem, and so if NVDA can find mathml it trys 
that. Admittedly this may be detrimental in some cases. Can you give a 
case where the mathml's speech for the math is significantly different 
than what you would expect? @jamie: Maybe we should provide a gui option 
in NVDA to allow alt-text in mathml to be rendered instead, for this 
type of special case? What do you guys think?

On 1/7/2016 2:17 PM, RD-Brown wrote:
>
> I work for Pearson Education’s accessibility team, and we have 
> identified an issue where NVDA does not read alt text descriptions for 
> MathML content used in online assessments with the Firefox browser We 
> are using alt text descriptions because accurate reading of MathML 
> content varies so widely depending on screen reader versions and the 
> browser being used
> (XML attached)
>
> NVDA_BugExampletxt 
> <https://githubcom/nvaccess/nvda/files/81757/NVDA_BugExampletxt>
> 1 Open the following link in Firefox: 
> uat-collegeboardtestnavcom/client/indexhtml#login?username=LGN557120611&password=MDYQWJ8G
> 2 On the Start Test Page, press B until the Start Test Now button is 
> selected, and press Enter
> 3 On the Section Start Page, press B until the Section Start button is 
> selected, and press Enter
> 4 On the Question Page, press H until the Question heading is selected
> 5 Press Down arrow
>
> • Expected: NVDA should read the following text:
> It is given that sine x equals, a where x is the radian measure of an 
> angle and
> • Actual: NVDA skips the MathML alt text and instead reads:
> It is given that where x is the radian measure of an angle and
> • Expected: On the following line, NVDA should read:
> the fraction pi over 2, is less than x, and x is less than pi
> • Actual: NVDA reads nothing
>
> Interestingly, MathML is read in the radio button choices shown on 
> this page, but again not the alt text which was coded for these items 
> Instead, the MathML is being read directly
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/nvaccess/nvda/issues/5665>.
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.drupalgardens.com>
Non-proffessional website. <http://derekriemer.pythonanywhere.com/personal>
Awesome little hand built weather app that rocks! 
<http://derekriemer.pythonanywhere.com/weather>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

**Comment by jamie** [comment]
Wrong @jamie :)

**Comment by derekriemer** [comment]
Woops. Sorry

On 1/7/2016 2:47 PM, Jamie Macey wrote:
>
> Wrong @jamie <https://github.com/jamie> :)
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/nvaccess/nvda/issues/5665#issuecomment-169816302>.
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.drupalgardens.com>
Non-proffessional website. <http://derekriemer.pythonanywhere.com/personal>
Awesome little hand built weather app that rocks! 
<http://derekriemer.pythonanywhere.com/weather>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

**Comment by RD-Brown** [comment]
<< I believe NVDA purposely ignores alt-text because it renders Methyl on
its own. >>



NVDA is not reading the MathML in the question however, only in the choice
interactions.  Also, when this item is viewed in Internet Explorer, NVDA
does read alt text in the question.



At the moment, because voicing and displaying of MathML in braille is such
a moving target, we feel that it is very important to provide an alt text
alternative.  Is it possible to read alt text when it is provided, and
directly render MathML when it is not?

**Comment by derekriemer** [comment]
That is not the default currently and is (I think) not available yet. I 
think it may be good to propose this as a change to NVDA (to allow an 
option to preference alt-text).

On 1/7/2016 4:02 PM, RD-Brown wrote:
> << I believe NVDA purposely ignores alt-text because it renders Methyl on
> its own. >>
>
>
>
> NVDA is not reading the MathML in the question however, only in the choice
> interactions. Also, when this item is viewed in Internet Explorer, NVDA
> does read alt text in the question.
>
>
>
> At the moment, because voicing and displaying of MathML in braille is such
> a moving target, we feel that it is very important to provide an alt text
> alternative. Is it possible to read alt text when it is provided, and
> directly render MathML when it is not?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/nvaccess/nvda/issues/5665#issuecomment-169834734>.
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.drupalgardens.com>
Non-proffessional website. <http://derekriemer.pythonanywhere.com/personal>
Awesome little hand built weather app that rocks! 
<http://derekriemer.pythonanywhere.com/weather>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

**Comment by jcsteh** [comment]
I do not believe preferring the alt text over the MathML is the correct behaviour. NVDA's MathML access provides a much better, more useful experience than alt text; i.e. it can be read in braille, it pauses appropriately, you can interactively explore the math, etc. If we prefer alt text, users will lose all of this whenever alt text is present. Furthermore, they won't necessarily understand *why* they're losing it; users can't be expected to understand what "alt text" is and when it is present.

The bigger question for me is why the math isn't being read for one of the expressions.

**Comment by RD-Brown** [comment]
If NVDA and other screen readers reliably provided all of the MathML features you listed, then I agree, alt text for MathML content would not be necessary.  However, since content is not being reliably rendered, and we have to somehow convey math concepts to screen reader users, alt text is currently necessary.

Would it be feasible to allow for a user preference that would enable/disable alt text presented with MathML content?

**Comment by RD-Brown** [comment]
Greetings,

Can NV Access give any sort of timeline as to when the issue above, or alternatively the direct reading of MathML, might be resolved?

**Comment by jcsteh** [comment]
For me, on the example page given, the direct math *is* read in Firefox. I hear:

"It is given that, sine, x, equals, a, where x is the radian measure of an angle and"

Do you have Design Science MathPlayer 4 installed as documented in the User Guide?

Note that NV Access will not commit to any timelines without a corporate support contract in place.

**Comment by RD-Brown** [comment]
James,

Thank you for your response.  To verify our results, I uninstalled MathPlayer, rebooted, then downloaded and installed a fresh copy.  Unfortunately, MathML is still being read incorrectly in Firefox.  I'm running Windows 7, along with the latest version of FF, 44.0.2, and still hear:
"It is given that   where x is the radian measure of an angle and "

Do you have any additional suggestions?

Dan

**Comment by jcsteh** [comment]
Can you please uninstall MathPlayer, completely delete its program 
directory, reboot and then reinstall? I've heard of bugs in the 
MathPlayer (un)installer which cause problems like this. Also, please be 
sure this is MathPlayer 4 beta, not MathPlayer 3.

**Comment by RD-Brown** [comment]
Dear James,

I performed the following steps:
* Uninstalled MathPlayer through Add/Remove programs.
* Deleted the Design Science folder under \Program Files (X86)
* Emptied Recycle Bin
* Rebooted
* Installed MathPlayer V4 (There's no mention I could find of a beta on the Design Science website, but this version is clearly labeled as V4.)

Unfortunately, MathML is still being read incorrectly as originally reported.

Pearson is very interested in resolving this issue ASAP.  We are impressed with NVDA's growing functionality, and have chosen it as the first working screen reader in our own internal accessible equation editor project.  We're happy to consider a corporate sponsorship, and I invite you to contact me by e-mail, dan.brown@pearson.com, so that we can discuss available options and sponsorship levels.  Unfortunately, the beginning of Spring assessments for our PARCC schools across several states in the U.S. is in two weeks, and it's doubtful we could initiate anything in that limited time.  Barring a solution, we will have to advise them soon that NVDA cannot be used as a screen reader while taking math assessments.

Please let me know of any additional suggestions you may have to resolve this problem.  I can also be reached via Skype under rdan.brown.

Dan
